### PR TITLE
Create TrackSetTableModel as base for Crate- & PlaylistTableModel

### PR DESCRIPTION
**Pull request by xeruf**

### CODE REVIEW

File: src/library/crate/cratetablemodel.h [hunk 1]
```diff
@@ -1,37 +1,30 @@
...
 
   public:
     CrateTableModel(QObject* parent, TrackCollectionManager* pTrackCollectionManager);
     ~CrateTableModel() final;
 
-    void selectCrate(
-        CrateId crateId = CrateId());
-    CrateId selectedCrate() const {
+    void selectCrate(CrateId crateId = CrateId());
+    [[nodiscard]] CrateId selectedCrate() const {
```
Comment (Holzhaus) [inline src/library/crate/cratetablemodel.h:14]: We don't use this anywhere else. Now that we switched to C++17 it might be a good idea to start using it for simple getter functions like this. @uklotzde @daschuer Opinions?

File: src/library/playlisttablemodel.cpp [hunk 2]
```diff
@@ -8,13 +8,14 @@
 
 #include "mixer/playermanager.h"
 
+/// \param preserveAll Whether to keep deleted tracks, default false
```
Comment (Holzhaus) [inline src/library/playlisttablemodel.cpp:11]: I think this should go in the header file. And if you use param, you should use `@` since that's already used in our code base.

File: src/library/tracksettablemodel.h [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include "library/basesqltablemodel.h"
+
+class TrackSetTableModel : public BaseSqlTableModel {
+    Q_OBJECT
+
+};
```
Comment (Holzhaus) [inline src/library/tracksettablemodel.h:8]: Why does this class exist if it doesn't define any methods?

File: src/library/crate/cratetablemodel.h (hunk 1 above)
Comment (uklotzde) [reply src/library/crate/cratetablemodel.h:14]: It improves safety.

On the other hand it introduces a lot of noise to the code and makes it harder to read. Functions return results for a reason and results should never be discarded. If the result is not mandatory it should be `void`, indicating that the "purpose" of this function is to cause a side-effect.

Personally I would instead like to enable a warning for **all** discarded return values by default. The caller should be responsible to explicitly **opt-out** for every discarded return value. Considering the amount of legacy code this might not be feasible.

File: src/library/playlisttablemodel.cpp (hunk 2 above)
Comment (xerus2000) [reply src/library/playlisttablemodel.cpp:11]: you're totally right - about @ vs / I simply wasn't sure, will adjust

File: src/library/tracksettablemodel.h (hunk 3 above)
Comment (xerus2000) [reply src/library/tracksettablemodel.h:8]: because it's a WIP ;)

Comment (Holzhaus) [reply src/library/tracksettablemodel.h:8]: Okay :smile:

File: src/library/crate/cratetablemodel.h (hunk 1 above)
Comment (xerus2000) [reply src/library/crate/cratetablemodel.h:14]: just followed the instructions of clang-tidy here - maybe we can configure that as well to fit your style?

Comment (uklotzde) [reply src/library/crate/cratetablemodel.h:15]: This is just a simple, idempotent getter method. I don't see any value in introducing this attribute here.

An agreed set of rules for clang-tidy would be preferable. Otherwise we will argue repeatedly where a certain style is appropriate. The same situation already applies to many other informal C++ conventions that are not checked by the compiler.

Comment (daschuer) [reply src/library/crate/cratetablemodel.h:15]: Reading here: https://en.cppreference.com/w/cpp/language/attributes/nodiscard
that new() has this attribute. It is clear that a bad thing = "memory leak" happens when the return value is discarded. This is obvious during writing the called function. 
I think it is reasonable to follow this pattern.  

The crucial point here is if, this applies to simple getter functions. If a user discards the value,
I guess the compiler will discard the whole call. So nothing bad happens form the called function perspective. However it is  most likely an issue from the caller perspective. 

I think I would not use it here in favor of less boilerplate.

File: src/library/basesqltablemodel.cpp [hunk 4]
```diff
@@ -38,18 +38,12 @@ const QString kEmptyString = QStringLiteral("");
...
-        const char* settingsNamespace)
-        : BaseTrackTableModel(
-                  settingsNamespace,
-                  pTrackCollectionManager,
-                  parent),
-          m_pTrackCollectionManager(pTrackCollectionManager),
-          m_database(pTrackCollectionManager->internalCollection()->database()),
-          m_bInitialized(false),
-          m_currentSearch(kEmptyString) {
+BaseSqlTableModel::BaseSqlTableModel(QObject* parent, TrackCollectionManager* pTrackCollectionManager, const char* settingsNamespace)
```
Comment (daschuer) [inline src/library/basesqltablemodel.cpp:41]: This line of code exceeds the GitHub limit. Please break.
Did clang-format unbreak it?  It can be tricked by manually breaking after (. 
I think the recent pre-commit hook can do it as well. 
Please verify the whole PR.

File: src/library/basetracktablemodel.h [hunk 5]
```diff
@@ -18,9 +18,9 @@ class BaseTrackTableModel : public QAbstractTableModel, public TrackModel {
 
   public:
     explicit BaseTrackTableModel(
-            const char* settingsNamespace,
+            QObject* parent,
             TrackCollectionManager* const pTrackCollectionManager,
-            QObject* parent);
+            const char* settingsNamespace);
```
Comment (daschuer) [inline src/library/basetracktablemodel.h:23]: If you like this can be a QString initialized by a QStringLiteral()
Since we store the value finally, you may use a value parameter + std::move() when passing.

File: src/library/playlisttablemodel.cpp [hunk 6]
```diff
@@ -11,10 +11,10 @@
 PlaylistTableModel::PlaylistTableModel(QObject* parent,
                                        TrackCollectionManager* pTrackCollectionManager,
                                        const char* settingsNamespace,
-                                       bool showAll)
-        : BaseSqlTableModel(parent, pTrackCollectionManager, settingsNamespace),
+                                       bool preserveAll)
+        : TrackSetTableModel(parent, pTrackCollectionManager, settingsNamespace),
           m_iPlaylistId(-1),
-          m_showAll(showAll) {
+          m_preserveAll(preserveAll) {
```
Comment (daschuer) [inline src/library/playlisttablemodel.cpp:17]: This was really unclear. 
How about: m_preserveHiddenTracks or invert m_removeHiddenTracks

**Review by daschuer — COMMENTED** [review]
Thank you for working on this. Here are some comments.

File: src/library/crate/cratetablemodel.h (hunk 1 above)
Comment (Holzhaus) [reply src/library/crate/cratetablemodel.h:15]: > The crucial point here is if, this applies to simple getter functions. If a user discards the value,
I guess the compiler will discard the whole call. So nothing bad happens form the called function perspective.

It will likely be discarded (hehe) by the compiler, but it clearly indicates bad code. I'd be in favor of using it.

Comment (uklotzde) [reply src/library/crate/cratetablemodel.h:15]: But using it only here and not everywhere else and especially in places where it actually would be more appropriate as Daniel explained is questionable.

Comment (uklotzde) [reply src/library/crate/cratetablemodel.h:15]: This is what `-Wunused-result` is for instead of polluting the code. The attribute is only appropriate where ignoring the return value is an error from the perspective of the callee, i.e. by causing memory leaks. Otherwise the function itself should not need to care by whom it is called and how the result is used or not.

File: src/library/basesqltablemodel.cpp (hunk 4 above)
Comment (xerus2000) [reply src/library/basesqltablemodel.cpp:41]: I tried to get the code somewhat consistent and readable, but yeah I also wasn't quite sure here. I'll adjust it.

File: src/library/playlisttablemodel.cpp (hunk 6 above)
Comment (Holzhaus) [reply src/library/playlisttablemodel.cpp:17]: If we change what it does, we can just rename it then.

Comment (xerus2000) [reply src/library/playlisttablemodel.cpp:17]: fair point ^^

File: src/library/basetracktablemodel.h (hunk 5 above)
Comment (xerus2000) [reply src/library/basetracktablemodel.h:23]: This would have to be changed in quite a few places, so I'd like to refrain from that for now

**Comment by xerus2000** [comment]
Since this is already quite big, I'd prefer to merge this as is and then actually implement TrackSetTableModel in a separate PR.

@Be-ing could you merge?

File: src/library/setlogfeature.cpp [hunk 7]
```diff
@@ -27,7 +27,7 @@ SetlogFeature::SetlogFeature(
                           nullptr,
                           pLibrary->trackCollections(),
                           "mixxx.db.model.setlog",
-                          /*show-all-tracks*/ true),
+                          true),
```
Comment (uklotzde) [inline src/library/setlogfeature.cpp:30]: Why did you remove this helpful comment? This comment reflected the fact that booleans are inappropriate as functions parameters in general. Such comments should not be removed without fixing the actual cause.

Especially in this case were setting it to false might cause the deletion of data from the database! ...which is bad and dangerous design decision imho, but that's another topic.

File: src/library/basesqltablemodel.cpp [hunk 8]
```diff
@@ -43,10 +43,7 @@ BaseSqlTableModel::BaseSqlTableModel(
         QObject* parent,
         TrackCollectionManager* pTrackCollectionManager,
         const char* settingsNamespace)
-        : BaseTrackTableModel(
-                  settingsNamespace,
-                  pTrackCollectionManager,
-                  parent),
+        : BaseTrackTableModel(parent, pTrackCollectionManager, settingsNamespace),
```
Comment (uklotzde) [inline src/library/basesqltablemodel.cpp:46]: Why did you reformat this code? Both formats are valid.

Keeping parameters of long parameter or argument lists on separate lines allows to easily review changes. In general long parameter lists should be avoided.

Comment (Holzhaus) [reply src/library/basesqltablemodel.cpp:46]: I agree, and unnecessary reformatting should generally be avoided. In this case we can still merge it either way because the resulting line is still short enough IMHO.

Comment (xerus2000) [reply src/library/basesqltablemodel.cpp:46]: Because I wanted to maintain a consistent style throughout the whole class hierarchy. And since it is merely passing all arguments up that are given to it, which are already listed on separate lines above, I don't think that it is worth to list them all on separate lines again.

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (xerus2000) [reply src/library/setlogfeature.cpp:30]: Because I changed the parameter name to be more descriptive. Imo you should consult your IDE for such information about parameters, not thrown-in comments.

Comment (xerus2000) [reply src/library/setlogfeature.cpp:30]: For example, Intellij/CLion actually prefixes values in such calls with their name:
image

Comment (uklotzde) [reply src/library/setlogfeature.cpp:30]: I disagree. The code must be comprehensible without an IDE. Please don't make such assumptions.

Comment (xerus2000) [reply src/library/setlogfeature.cpp:30]: I think that is a subjective general thing, so I'd like to get some more opinions. @Be-ing @daschuer @Holzhaus :)

@uklotzde would you then expect any boolean parameter to have such a comment? I feel like this ruins the point of sensible parameter naming.

Comment (xerus2000) [reply src/library/setlogfeature.cpp]: I'll be adding it back for now so we can merge, but I'd still like to talk about this.

**Comment by xerus2000** [comment]
I hope we can finally merge now :)

Follow-up tasks:
- [ ] [Replace char* parameter with QString](https://github.com/mixxxdj/mixxx/pull/2717#discussion_r416939871)
- [ ] Refactor methods into TrackSetTableModel

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (uklotzde) [reply src/library/setlogfeature.cpp]: Boolean parameters don't read well at the call site. That's the point why we recommend to use enums instead in the future.

But there is no excuse for making the code worse without providing an alternative. This is not negotiable.

Comment (uklotzde) [reply src/library/setlogfeature.cpp]: @xerus2000 I already explained in detail why I think that the comment for this *particular* boolean parameter matters:

> Especially in this case were setting it to false might cause the deletion of data from the database!

**Review by Holzhaus — APPROVED** [review]
I would merge but due to the discussion in #2699 I'll wait until @uklotzde has re-reviewed.

**Comment by xerus2000** [comment]
@Holzhaus the code changes concerning his comments are both resolved :)

File: src/library/setlogfeature.cpp (hunk 7 above)
Comment (xerus2000) [reply src/library/setlogfeature.cpp]: Moved: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/About.20coding.20style/near/198313605
